### ISSUE TITLE
Python 2.7 transient failure on 0.4.5

### ISSUE THREAD

**Issue by domenkozar** [python-greenlet/greenlet#85]
I've been seeing some transient failures like:

```

======================================================================
FAIL: test_threaded_adv_leak (tests.test_leaks.ArgRefcountTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/nix-build-python2.7-greenlet-0.4.5.drv-0/greenlet-0.4.5/tests/test_leaks.py", line 70, in test_threaded_adv_leak
    self.assertTrue(g() is None)
AssertionError: False is not true
```

See whole build failure at http://hydra.nixos.org/build/22202676/nixlog/1/raw

**Comment by snaury** [comment]
Thanks, I'm aware of this failure (very annoying when it triggers, but so far I've only seen this on 2.6 and on travis). The test was designed to trigger 100% of the time when the bug was present, but these transient failures probably mean `t.join()` is not enough to reliably ensure gc happens when all of relevant objects are collectable, but so far I haven't been able to fix it. The biggest problem is that I haven't been able to reproduce this on any of my machines at all so far, so I won't know when it's fixed or not. :( So far I've tried calling `gc.collect` multiple times, but it obviously didn't help.

If you can reproduce this, could you please check if adding something like `time.sleep(1)` after the `for i in range` loop fixes these transient errors?

**Comment by domenkozar** [comment]
I can't reproduce it, but our Hydra build farm has 50 cores and heavy IO load. It only happen there occasionally. 

Maybe try introducing artificial IO load on your machine and run the tests and see if that breaks it?

**Comment by snaury** [comment]
Oh, I actually finally found a way how to make it fail, I just added this in my tree:

```
        def test_threaded_adv_leak_insane(self):
            for i in xrange(1000000):
                self.test_threaded_adv_leak()
```

Which made it easier to reproduce. Unfortunately my first attempts made this problem less likely to occur, but didn't eliminate it completely...

**Comment by snaury** [comment]
Hopefully fixed in #86. It's still not 100% (because there's no cross-platform way to find when the thread is destroyed in Python, only when Python code on that thread has stopped executing, which is too early). But on my laptop I did 2 million iteration (~3 hours) and didn't have any failures, so I hope chance of hitting that is much lower now.
